### DISCUSSION TITLE
Avoiding global variables in functions

### DISCUSSION THREAD

**Discussion by richardreeve** [IBAHCM/RPiR#443] [Announcements]
Hi @IBAHCM/rpir21 

Previously we advised you to put a short `if ()` statement in your helper files that would look in your functions for global variables and generate errors to stop you using them. Like so:

```R
library(codetools)
if (length(findGlobals(timestep_deterministic_SIS,
                       merge = FALSE)$variables) != 0) {
  stop(
    "Function timestep_deterministic_SIS() may not use global variable(s): ",
    findGlobals(timestep_deterministic_SIS, merge = FALSE)$variables
  )
}
```

We've now written a short function in RPiR for you that does this for you - `assert_no_globals()`. It does exactly the same as the `if ()` statement, but avoids you having to write so much duplicated code:

```
RPiR::assert_no_globals(timestep_deterministic_SIS)
```

This should go into the bottom of every file with a function definition in the `R` folder of your package and it will stop you running `document()` or `install()` if your package functions use global variables.

**Comment by 2700638AichiMkunde** [comment]
@richardreeve @soniamitchell 

do I create that one as a function and save with other function scripts along with?

this assert_no_globals() function?

**Reply by JumaMinya** [reply]
@2700638AichiMkunde 
The phrase
**RPiR::assert_no_globals(timestep_deterministic_SIS)** 

replaces
if (length(findGlobals(timestep_deterministic_SIS,
                       merge = FALSE)$variables) != 0) {
  stop(
    "Function timestep_deterministic_SIS() may not use global variable(s): ",
    findGlobals(timestep_deterministic_SIS, merge = FALSE)$variables
  )
}
that we have been using.

So we only use 
RPiR::assert_no_globals(timestep_deterministic_SIS)

**Reply by 2700638AichiMkunde** [reply]
Thank you @JumaMinya

**Comment by richardreeve** [comment]
@SBOHVM/rpir22 - just tagging you to say this is the global variable solution that I mentioned.
